rops_check.py

```diff
@@ -0,0 +1,149 @@
+
+# The percentage threshold for the number of ROP gadgets to be accepted.
+PERCENTAGE_THRESHOLD = 10
+
+DESCRIPTION = "This script compares the number of ROP (Return-Oriented      \
+    Programming) gadgets in two different builds from two branches instances\
+    of a Git repository. The script calls the ROPGadget* tool to do         \
+    the measurements and outputs the difference in the number of ROP gadgets\
+    between the two provide branches.                                       \
+    * github.com/JonathanSalwan/ROPgadget"
+
+def parse_args():
+    """
+    This function parses command-line arguments using the argparse module. It
+    sets up the argument parser with the specified options and returns the
+    parsed arguments.
+    """
+    parser = argparse.ArgumentParser(description=DESCRIPTION)
+    parser.add_argument("-s", "--source-branch", required=False,
+                        help="Name of source branch to compare against (default: current branch)")
+    parser.add_argument("-t", "--target-branch", required=False,
+                        help="Name of target branch to compare (default: main)")
+    parser.add_argument("-b", "--build-cmd", required=True, help="Command to build the repo")
+    parser.add_argument("-x", "--exe-path", required=True,
+                        help="Path to the executable to measure ROP gadgets in")
+    parser.add_argument("-a", "--exe-arch", choices=['arm','arm64', 'riscv'], required=True,
+                        help="Architecture of the executable to measure ROP gadgets in")
+    parser.add_argument("-p", "--pct", required=False,
+                        help="Percentage threshold for the number of ROP gadgets to be accepted \
+                            (default: 10%%)")
+
+    args = parser.parse_args()
+    return args
+
+def run_cmd(cmd):
+    """
+    This function takes a command as input and executes it using the subprocess
+    module. It captures the output (stdout) and error (stderr) streams, as well
+    as the return code.
+    """
+    # Run the command using the subprocess module and capture the output, error, and return code.
+    with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True,
+                          text=True) as process:
+
+        # Wait for the command to complete and retrieve the output and error streams.
+        output, error = process.communicate()
+
+        # Get the return code of the command.
+        return_code = process.returncode
+
+    # Return the output, error, and return code as a tuple.
+    return output, error, return_code
+
+def measure_rop_gadgets(exe_path, exe_arch):
+    """
+    This function takes a path to a binary as input and uses the ROPGadget tool
+    to measure the number of ROP gadgets in the binary.
+    """
+    # Run the ROPGadget tool on the binary and capture the output and error streams.
+    cmd = f'ROPgadget --binary {exe_path} --rawEndian little --rawMode 64 --rawArch {exe_arch}'
```

Comment:
I've been experimenting with `ROPgadget` and noticed we can pass an ELF file directly to `--binary` instead of a raw binary. So, I don't think we need the raw* arguments, because we will always have the ELF files.

Also, if you pass a raw binary we lose address information which is later useful to understand where exactly in the code the gadgets are using `addr2line`.

Reply:
Maybe we need to clarify in line 42 the executable is not a raw binary but actually an executable file format.

Reply:
I changed to only consider that only ELF files are passed to the script. It's only to be used with such purpose so I can assume that only files will be passed.